volatility/framework/symbols/windows/extensions/__init__.py

```diff
@@ -872,3 +874,304 @@ def get_wait_reason(self) -> str:
+        """Get the available pages that correspond to a cached file.
+
+        The tuples generated are (physical_offset, file_offset, page_size).
+        """
+        symbol_table_name = self.get_symbol_table_name()
+        mmpte_type = self._context.symbol_space.get_type(symbol_table_name + constants.BANG + "_MMPTE")
+        mmpte_size = mmpte_type.size
+        subsection = self.get_subsection()
+        is_64bit = symbols.symbol_table_is_64bit(self._context, symbol_table_name)
+        is_pae = self._context.layers[self.vol.layer_name].metadata.get("pae", False)
+
+        # This is a null-terminated single-linked list.
+        while subsection != 0:
+            try:
+                if subsection.ControlArea != self.vol.offset:
+                    break
+            except exceptions.InvalidAddressException:
+                break
+
+            # The offset into the file is stored implicitly based on the PTE location within the Subsection.
+            starting_sector = subsection.StartingSector
+            subsection_offset = starting_sector * 0x200
+
+            # Similar to the check in is_valid(), make sure the SubsectionBase is not page aligned.
+            #if subsection.SubsectionBase & self.PAGE_MASK == 0:
+            #    break
+
+            ptecount = 0
+            while ptecount < subsection.PtesInSubsection:
+                pte_offset = subsection.SubsectionBase + (mmpte_size * ptecount)
+                file_offset = subsection_offset + ptecount * 0x1000
+
+                try:
+                    mmpte = self.get_pte(pte_offset)
+                except exceptions.InvalidAddressException:
+                    ptecount += 1
+                    continue
+
+                # First we check if the entry is valid. If so, then we get the physical offset.
+                # The valid entries are actually handled by the hardware.
+                if mmpte.u.Hard.Valid == 1:
+                    physoffset = mmpte.u.Hard.PageFrameNumber << 12
+                    yield physoffset, file_offset, self.PAGE_SIZE
+
+                elif mmpte.u.Soft.Prototype == 1:
+                    if not is_64bit and not is_pae:
+                        subsection_offset = ((mmpte.u.Subsect.SubsectionAddressHigh << 7) | (
+                                mmpte.u.Subsect.SubsectionAddressLow << 3))
+
+                # If the entry is not a valid physical address then see if it is in transition.
+                elif mmpte.u.Trans.Transition == 1:
+                    physoffset = mmpte.u.Trans.PageFrameNumber << 12
+                    yield physoffset, file_offset, self.PAGE_SIZE
+
+                # Go to the next PTE entry
+                ptecount += 1
+
+            # Go to the next Subsection in the single-linked list
+            subsection = subsection.NextSubsection
+
+class VACB(objects.StructType):
+    """A class for _VACB structures"""
+
+    FILEOFFSET_MASK = 0xFFFFFFFFFFFF0000
```

Comment:
Still seems kinda hard-codey.  Can we calculate this in any way?  Are we certain we'll never get something longer than that?  Also not a show stopper but kinda prefer avoiding specific values particularly for masking things, etc.